### DISCUSSION TITLE
New default settings for 2.9.0

### DISCUSSION THREAD

**Discussion by lubber-de** [fomantic/Fomantic-UI#1745] [General]
### Which settings should change in 2.9.0 ?

I would like to collect suggestions for changes of default values for 2.9.0 (some are already prepared in [PRs](https://github.com/fomantic/Fomantic-UI/milestone/27))

While working with FUI, sometimes you stumble upon behaviors which seems to be wrong or annoying and you always use a different setting for your usage. While you of course can override them globally, i think we should change some values in the core directly. As it is a breaking change this has to wait until 2.9.0

Whenever you suggest some change, please also add a little explanation why you think that value should change.
By using the Emoji-Reactions on each suggestion we can see how many people agree to that

Let me start 🙂 

#### errorFocus
current value: `false`
new value: `true`
PR: #2041 

The new default will always focus the first error field of a form which is more logical. To adjust a form field content you almost always have to click into that field with the current Logic

#### forceSelection

current value: `true`
new value: `false`
PR: #1747

##### suggested by @larsbo in #1742:

If you click into a dropdownfield and just click somewhere else again, the first entry of the dropdown is always selected which is most probably not always the desired behavior. A given `required` attribute from a to be converted `select` should automatically set `forceSelection` to `true` again.

**Comment by maidzen** [comment]
### Dropdown
**fullTextSearch**
current value: `false`
new value: `'exact'`
PR #2042 

used fomantic in a few projects with different people 90% of the time a search Dropdown was used "exact" was the value wanted
Edit:
For example often we use multiple words in options like "Microsoft Word", "Adobe Photoshop"
"fullTextSearch: true" will not return "Microsoft Word" when the user searchinput is "Word" this is in many realworld examples an unwanted behavior.
Name Selection.. Option "John Doe" you want to be able to search for John and for Doe... and get this option as result.

**Comment by lubber-de** [comment]
### Calendar
#### constantHeight
current value: `true`
new value: `false`

Whenever the first of a month is shown within the first 5 columns of a calendar, then, according to the months amount of days, a full adjacent row is added to the calendar, which is unnecessary in most cases, so i suggest to turn this setting off by default

**Reply by lubber-de** [reply]
We will not change it. Reason here
https://github.com/fomantic/Fomantic-UI/pull/2043#issuecomment-895853437

**Comment by ko2in** [comment]
### Calendar
**monthFirst**

Current value: `true`
New value: `false`

The standard date format widely used is `dd-mm-yyyy`. I always used `dd-mm-yyyy` or `yyyy-mm-dd`. So, in my opinion, the month shouldn't be first by default. And we've several issues reported in FUI repo regarding the date starting with month and fixed them in the past, since JavaScript also has inconsistent results without specific fix while parsing the date starting with the month for different browsers. So, I think the value `false` would be less error prone for some default situations.

### Modal

**autofocus**

Current value: `true`
New value: `false`

This option tries to focus the first field of form if the modal has a form. And it's annoying when the form contains the dropdown as the first field or form with only dropdown. It makes the dropdown to be opened automatically because of focus event. And also, if the field has some special logic binds with focus event, this would also trigger automatically which is unexpected in most situations. And modal and form do not belong together all the time, so I think it make sense to lets the user decide whether to enable focus when the form is required and wants the field to be focused.

I disabled this option for all modals in my project so, I needed to provide this option all the time. Most of my forms load dynamically within the modal via Ajax, and I can't predict whether the form would have dropdown as first field. So, disabled this option would help to avoid this situations.

**Reply by lubber-de** [reply]
Regarding autofocus:  I agree. But i suggested the opposite for the `errorFocus` in form above... Of course ,if the first error field in a form would be a dropdown as well we have the same situation like in modal now... So should we keep the errorFocus as it is then? Or do you think its independent?  I am unsure 🤔

**Reply by ko2in** [reply]
`errorFocus` is fine as you suggested since the user already expected to interact with form again if the form is not valid. So, if the first error field is dropdown, and it automatically opens then, it makes sense in user experience perspective.

But for the modal, if the user is willing to open it, what he mostly expects is the modal to open and display the content inside it. Not to interact immediately with the content inside it.

**Reply by jamessampford** [reply]
Just a thought, but it could always be that the default for the dropdown for **showOnFocus** is instead changed from `true` to `false` (https://jsfiddle.net/0pyja8oz/) as this will stop the dropdown from opening but allowing it to retain focus and would be better in terms of accessibility (in a manner of speaking)

I don't believe `<select>` tags normally open when they gain focus (at least with autofocus) anyway (only when the enter key is hit does it open) and using the arrow keys goes through the options without it opening at all. Of course if it is clicked or tapped on, it should open...

**Reply by ko2in** [reply]
Of course, `showOnFocus: false` can prevent opening the dropdown. But we're talking about the default behavior with common usages. So, opening modal doesn't primarily intend to interact with form elements as my experiences with FUI in my projects.

So, it shouldn't be focused by default, it only needs when I've form fields inside the modal and my UI implementation requires focus for some user experience in rare situations.

**Reply by jamessampford** [reply]
As always, it depends on use cases as person A’s needs will differ from B’s as well as C’s - with all 3 being equal and valid - changing defaults will always cause work for someone whom is already using said default

**Reply by ko2in** [reply]
True and that's why @lubber-de started this discussion to get feedback from FUI users to change the default settings base on their experiences and common usages. 

If you think the dropdown should have `showOnFocus: false` as default, then please write your discussion thread instead of replying here. I didn't give any suggestion about dropdown in my discussion yet, we're talking about the modal's behavior. So, we're slightly out of topic now.

**Reply by lubber-de** [reply]
After implementing the new modal config templates and the related examples like "prompt" i now disagree to change the default for modals autoFocus #2044 . I am also preparing some changes for modal which only makes sense when autofocus is enabled, so i second the suggestion to change the default for dropdowns showOnfocus (which is the more annoying behavior in comparison.
Modal saves and restores the previous focussed element anyway

**Reply by lubber-de** [reply]
> So, it shouldn't be focused by default, it only needs when I've form fields inside the modal and my UI implementation requires focus for some user experience in rare situations.

In Addition, it's not only the form elements (input, textarea) which are affected by autofocus but also any button in the action bar.

**Comment by lubber-de** [comment]
### API
#### rawResponse
current value: `false`
new value: `true`
PR: #2230 

The new default will not force convert JSON Arrays to JSON Objects anymore when provided to the onResponse event handler.
JSON array is a valid JSON type
